File: frontend/src/services/agents/coordinatorAgent.ts
/**
 * Coordinator Agent
 *   Input:      AirAgentResult + WaterAgentResult + WasteAgentResult (any may be missing)
 *   Processing: cross-signal reasoning -> risk aggregation -> priority identification
 *   Output:     CoordinatorResult
 *
 * It only ever sees specialist reports, never raw sensor/provider data.
 */
import { clamp, formatNumber, joinAnd, riskLevelFor, roundHalfUp } from '../../lib/risk';
import type {
  ContributingFactor,
  CoordinatorInput,
  CoordinatorResult,
  Recommendation,
  RiskLevel,
  SignalContribution,
  SpecialistAgentId,
  SpecialistResult,
} from '../../types/agents';
import { AppError } from '../errors';
import type { Agent, AgentTrace } from './agent';

export const COORDINATOR_WEIGHTS: Record<SpecialistAgentId, number> = { air: 0.38, water: 0.27, waste: 0.35 };
const CORRELATION_BOOST = 0.03;
const MISSING_CONFIDENCE_PENALTY = 0.1;
const MAX_FACTORS = 3;
const ORDER: SpecialistAgentId[] = ['air', 'water', 'waste'];

const HIGH_PHRASE: Record<SpecialistAgentId, string> = { air: 'air pollution', water: 'water-quality degradation', waste: 'litter density' };
const MODERATE_PHRASE: Record<SpecialistAgentId, string> = {
  air: 'air quality shows moderate deterioration',
  water: 'water-quality indicators show moderate degradation',
  waste: 'litter levels are moderately elevated',
};
const LOW_PHRASE: Record<SpecialistAgentId, string> = {
  air: 'air quality remains within acceptable limits',
  water: 'water-quality indicators remain within acceptable limits',
  waste: 'litter levels remain low',
};

const capitalise = (text: string) => text.charAt(0).toUpperCase() + text.slice(1);

type Candidate = [number, Omit<Recommendation, 'priority' | 'urgency'>];

export class CoordinatorAgent implements Agent<CoordinatorInput, CoordinatorResult> {
  readonly id = 'coordinator' as const;
  readonly name = 'Coordinator Agent';

  async run(input: CoordinatorInput, trace: AgentTrace): Promise<CoordinatorResult> {
    const reports = ORDER.map((id) => [id, input[id]] as const).filter(
      (entry): entry is readonly [SpecialistAgentId, SpecialistResult] => entry[1] !== null,
    );
    if (!reports.length) throw new AppError('ANALYSIS_FAILED', 'Coordinator received no specialist reports to assess.');
    const missing = ORDER.filter((id) => input[id] === null);
    trace.log(`Received reports from ${reports.length} specialist agent${reports.length !== 1 ? 's' : ''}`);

    const weightTotal = reports.reduce((sum, [id]) => sum + COORDINATOR_WEIGHTS[id], 0);
    const contributions: SignalContribution[] = reports
      .map(([id, report]) => ({
        agent: id,
        riskLevel: report.riskLevel,
        riskScore: report.riskScore,
        weight: roundHalfUp(COORDINATOR_WEIGHTS[id] / weightTotal, 3),
        contribution: roundHalfUp((COORDINATOR_WEIGHTS[id] / weightTotal) * report.riskScore, 3),
      }))
      .sort((a, b) => b.contribution - a.contribution);
    const baseScore = reports.reduce((sum, [id, report]) => sum + (COORDINATOR_WEIGHTS[id] / weightTotal) * report.riskScore, 0);

    trace.log('Performing cross-signal reasoning...');
    const levelOf = (id: SpecialistAgentId): RiskLevel | undefined => input[id]?.riskLevel;
    const highs = ORDER.filter((id) => levelOf(id) === 'HIGH');
    const moderates = ORDER.filter((id) => levelOf(id) === 'MODERATE');
    const lows = ORDER.filter((id) => levelOf(id) === 'LOW');

    const adjustment = highs.length >= 2 ? CORRELATION_BOOST * (highs.length - 1) : 0;
    const overall = roundHalfUp(clamp(baseScore + adjustment));
    const overallLevel = riskLevelFor(overall);

    const insights = this.insights(input, highs, adjustment, missing);
    const rawConfidence = reports.reduce((sum, [id, report]) => sum + (COORDINATOR_WEIGHTS[id] / weightTotal) * report.confidence, 0);
    const confidence = roundHalfUp(clamp(rawConfidence - MISSING_CONFIDENCE_PENALTY * missing.length));
    trace.log(`Risk aggregation · ${(overall * 100).toFixed(0)}% overall (${overallLevel})`);

    const factors: ContributingFactor[] = [];
    for (const c of contributions) {
      const report = input[c.agent];
      if (report && report.findings.length && report.riskLevel !== 'LOW') {
        factors.push({ agent: c.agent, label: report.findings[0].label, detail: report.findings[0].detail });
      }
    }

    const recommendations = this.recommendations(input, overall, overallLevel);
    trace.log(`Identified ${recommendations.length} prioritized actions`);
    trace.log('Environmental risk assessment completed');

    return {
      agent: 'coordinator',
      location: input.location,
      overallRiskLevel: overallLevel,
      overallScore: overall,
      confidence,
      reasoning: this.reasoning(overallLevel, highs, moderates, lows, missing),
      crossSignalInsights: insights,
      contributingFactors: factors.slice(0, MAX_FACTORS),
      contributions,
      crossSignalAdjustment: roundHalfUp(adjustment, 3),
      dominantAgents: highs.length ? highs : moderates.length ? [contributions[0].agent] : [],
      inputsReceived: reports.map(([id]) => id),
      missingInputs: missing,
      recommendations,
      timestamp: new Date().toISOString(),
    };
  }

  private reasoning(level: RiskLevel, highs: SpecialistAgentId[], moderates: SpecialistAgentId[], lows: SpecialistAgentId[], missing: SpecialistAgentId[]): string {
    const elevated = highs.length + moderates.length;
    const opening =
      level === 'HIGH'
        ? elevated >= 2
          ? 'Multiple environmental indicators show elevated risk.'
          : 'A single dominant indicator is driving elevated environmental risk.'
        : level === 'MODERATE'
          ? 'Environmental indicators show moderate overall risk.'
          : 'Environmental indicators are largely within acceptable ranges.';

    let body: string;
    if (highs.length) {
      const verb = highs.length > 1 ? 'are the dominant risk factors' : 'is the dominant risk factor';
      const tail = moderates.length ? moderates.map((id) => MODERATE_PHRASE[id]) : lows.map((id) => LOW_PHRASE[id]);
      body = `${capitalise(joinAnd(highs.map((id) => HIGH_PHRASE[id])))} ${verb}${tail.length ? `, while ${joinAnd(tail)}.` : '.'}`;
    } else if (moderates.length) {
      const verb = moderates.length > 1 ? 'are the main concerns' : 'is the main concern';
      const tail = lows.map((id) => LOW_PHRASE[id]);
      body = `${capitalise(joinAnd(moderates.map((id) => HIGH_PHRASE[id])))} ${verb}${tail.length ? `, while ${joinAnd(tail)}.` : '.'}`;
    } else {
      body = 'No specialist agent reported indicators above guideline thresholds.';
    }

    let text = `${opening} ${body}`;
    if (missing.length) text += ` The ${joinAnd(missing)} assessment is unavailable, so confidence is reduced.`;
    return text;
  }

  private insights(input: CoordinatorInput, highs: SpecialistAgentId[], adjustment: number, missing: SpecialistAgentId[]): string[] {
    const insights: string[] = [];
    const { air, water, waste } = input;
    if (adjustment > 0) {
      insights.push(`Correlated high-risk signals from the ${joinAnd(highs)} agents add ${(adjustment * 100).toFixed(0)} points to the aggregate score.`);
    }
    if (waste && water && waste.riskLevel !== 'LOW' && water.turbidity !== null && water.turbidity > 5) {
      insights.push('Litter density alongside elevated turbidity suggests surface runoff may be carrying waste into the nearby water body.');
    }
    if (air && waste && air.riskLevel === 'HIGH' && waste.counts.plastic >= 8) {
      insights.push('High PM2.5 near a plastic-heavy litter zone warrants a check for open waste burning.');
    }
    if (water && water.sensorStatus === 'degraded') {
      insights.push('Water assessment relies on partial sensor data; one or more probes are not reporting.');
    }
    if (missing.length) {
      insights.push(`No report received from the ${joinAnd(missing)} agent; its weight was redistributed.`);
    }
    return insights;
  }

  private recommendations(input: CoordinatorInput, overall: number, overallLevel: RiskLevel): Recommendation[] {
    const { air, water, waste } = input;
    const candidates: Candidate[] = [];

    if (waste && waste.riskLevel !== 'LOW') {
      candidates.push([waste.riskScore, {
        id: 'inspect-waste', title: 'Inspect waste accumulation zone', agent: 'waste', actionLabel: 'Dispatch crew',
        explanation: `${waste.totalObjects} objects detected at ${waste.sourceName}, ${waste.counts.plastic} of them plastic. Dispatch a sanitation crew and trace the source.`,
      }]);
    }

    if (water && water.sensorStatus === 'degraded') {
      candidates.push([Math.max(water.riskScore, 0.5) * 0.95, {
        id: 'restore-sensor', title: 'Restore water sensor telemetry', agent: 'water', actionLabel: 'Open ticket',
        explanation: `${water.sensorName} is reporting partial data. Inspect probes and connectivity on ${water.sensorId}.`,
      }]);
    } else if (water && water.riskLevel !== 'LOW') {
      const turbidity = water.turbidity !== null && water.turbidity > 5
        ? `Turbidity at ${formatNumber(water.turbidity)} NTU exceeds the 5 NTU permissible limit`
        : 'Water indicators are drifting from safe ranges';
      candidates.push([water.riskScore * 0.95, {
        id: 'investigate-water', title: 'Investigate nearby water source', agent: 'water', actionLabel: 'Schedule sampling',
        explanation: `${turbidity} at ${water.sensorName}. Collect a grab sample for lab validation.`,
      }]);
    }

    if (overallLevel !== 'LOW') {
      candidates.push([overall * 0.7, {
        id: 'increase-monitoring', title: 'Increase environmental monitoring frequency', agent: 'all', actionLabel: 'Update cadence',
        explanation: 'Raise sampling cadence from 60 to 15 minutes across all agents for the next 24 hours to confirm the trend.',
      }]);
    }

    if (air && air.riskLevel === 'HIGH') {
      const pm = air.pm25 !== null
        ? `PM2.5 at ${formatNumber(air.pm25)} µg/m³ is ${(air.pm25 / 15).toFixed(1)}× the WHO guideline`
        : 'Air pollution is at a high-risk level';
      candidates.push([air.riskScore * 0.65, {
        id: 'air-advisory', title: 'Issue air-quality health advisory', agent: 'air', actionLabel: 'Draft advisory',
        explanation: `${pm}. Notify sensitive groups near ${air.stationName} and review traffic controls.`,
      }]);
    } else if (air && air.riskLevel === 'MODERATE') {
      candidates.push([air.riskScore * 0.6, {
        id: 'air-watch', title: 'Monitor air-quality trend', agent: 'air', actionLabel: 'Set alert',
        explanation: `Air quality at ${air.stationName} is moderate. Watch for sustained increases over the next 12 hours.`,
      }]);
    }

    if (!candidates.length) {
      candidates.push([0.1, {
        id: 'routine', title: 'Maintain routine monitoring', agent: 'all', actionLabel: 'Acknowledge',
        explanation: 'All indicators are within acceptable limits. Continue the standard sampling schedule.',
      }]);
    }

    return [...candidates]
      .sort((a, b) => b[0] - a[0])
      .map(([score, data], index) => ({
        ...data,
        priority: index + 1,
        urgency: index === 0 && score >= 0.6 ? 'Immediate' : score >= 0.45 ? 'Within 24h' : 'Routine',
      }));
  }
}

File: backend/agents/coordinator_agent.py
"""Coordinator Agent.

Input:      AirAgentResult, WaterAgentResult, WasteAgentResult (any may be missing)
Processing: cross-signal reasoning -> risk aggregation -> priority identification
Output:     CoordinatorResult (overall risk, reasons, recommended actions)

The coordinator never touches raw provider data — only specialist reports.
"""

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Dict, List, Optional, Tuple, Union

from agents.base import Agent, AgentTrace
from core.errors import AnalysisFailedError
from core.risk import clamp, format_number, join_and, risk_level_for, round_half_up
from schemas import (
    AirAgentResult,
    ContributingFactor,
    CoordinatorResult,
    Recommendation,
    SignalContribution,
    WasteAgentResult,
    WaterAgentResult,
)

SpecialistResult = Union[AirAgentResult, WaterAgentResult, WasteAgentResult]

WEIGHTS: Dict[str, float] = {"air": 0.38, "water": 0.27, "waste": 0.35}
CORRELATION_BOOST = 0.03  # per additional HIGH specialist signal
MISSING_CONFIDENCE_PENALTY = 0.1
MAX_FACTORS = 3

AGENT_NAMES = {"air": "air", "water": "water", "waste": "waste"}
HIGH_PHRASE = {"air": "air pollution", "water": "water-quality degradation", "waste": "litter density"}
MODERATE_PHRASE = {
    "air": "air quality shows moderate deterioration",
    "water": "water-quality indicators show moderate degradation",
    "waste": "litter levels are moderately elevated",
}
LOW_PHRASE = {
    "air": "air quality remains within acceptable limits",
    "water": "water-quality indicators remain within acceptable limits",
    "waste": "litter levels remain low",
}


@dataclass(frozen=True)
class CoordinatorInput:
    location: str
    air: Optional[AirAgentResult] = None
    water: Optional[WaterAgentResult] = None
    waste: Optional[WasteAgentResult] = None


def _capitalise(text: str) -> str:
    return text[:1].upper() + text[1:]


class CoordinatorAgent(Agent[CoordinatorInput, CoordinatorResult]):
    agent_id = "coordinator"
    name = "Coordinator Agent"

    def run(self, payload: CoordinatorInput, trace: AgentTrace) -> CoordinatorResult:
        reports: List[Tuple[str, SpecialistResult]] = [
            (agent_id, report)
            for agent_id, report in (("air", payload.air), ("water", payload.water), ("waste", payload.waste))
            if report is not None
        ]
        if not reports:
            raise AnalysisFailedError("Coordinator received no specialist reports to assess.")
        missing = [a for a in ("air", "water", "waste") if a not in {r[0] for r in reports}]
        trace.log(f"Received reports from {len(reports)} specialist agent{'s' if len(reports) != 1 else ''}")

        # --- risk aggregation
        weight_total = sum(WEIGHTS[a] for a, _ in reports)
        contributions = [
            SignalContribution(
                agent=agent_id,  # type: ignore[arg-type]
                risk_level=report.risk_level,
                risk_score=report.risk_score,
                weight=round_half_up(WEIGHTS[agent_id] / weight_total, 3),
                contribution=round_half_up(WEIGHTS[agent_id] / weight_total * report.risk_score, 3),
            )
            for agent_id, report in reports
        ]
        contributions.sort(key=lambda c: c.contribution, reverse=True)
        base_score = sum(WEIGHTS[a] / weight_total * r.risk_score for a, r in reports)

        trace.log("Performing cross-signal reasoning...")
        levels = {agent_id: report.risk_level for agent_id, report in reports}
        highs = [a for a in ("air", "water", "waste") if levels.get(a) == "HIGH"]
        moderates = [a for a in ("air", "water", "waste") if levels.get(a) == "MODERATE"]
        lows = [a for a in ("air", "water", "waste") if levels.get(a) == "LOW"]

        adjustment = CORRELATION_BOOST * (len(highs) - 1) if len(highs) >= 2 else 0.0
        overall = round_half_up(clamp(base_score + adjustment))
        overall_level = risk_level_for(overall)

        insights = self._insights(payload, highs, adjustment, missing)
        confidence = sum(WEIGHTS[a] / weight_total * r.confidence for a, r in reports)
        confidence = round_half_up(clamp(confidence - MISSING_CONFIDENCE_PENALTY * len(missing)))
        trace.log(f"Risk aggregation · {overall * 100:.0f}% overall ({overall_level})")

        factors: List[ContributingFactor] = []
        report_map = dict(reports)
        for c in contributions:
            report = report_map[c.agent]
            if report.findings and report.risk_level != "LOW":
                top = report.findings[0]
                factors.append(ContributingFactor(agent=c.agent, label=top.label, detail=top.detail))
        factors = factors[:MAX_FACTORS]

        recommendations = self._recommendations(payload, overall, overall_level)
        trace.log(f"Identified {len(recommendations)} prioritized actions")
        limitations = self._data_limitations(payload, missing)
        trace.log("Final environmental risk generated")

        dominant = highs or ([contributions[0].agent] if moderates else [])
        return CoordinatorResult(
            location=payload.location,
            overall_risk_level=overall_level,
            overall_score=overall,
            confidence=confidence,
            reasoning=self._reasoning(overall_level, highs, moderates, lows, missing),
            cross_signal_insights=insights,
            contributing_factors=factors,
            contributions=contributions,
            cross_signal_adjustment=round_half_up(adjustment, 3),
            dominant_agents=dominant,  # type: ignore[arg-type]
            inputs_received=[a for a, _ in reports],  # type: ignore[misc]
            missing_inputs=missing,  # type: ignore[arg-type]
            recommendations=recommendations,
            data_limitations=limitations,
            timestamp=datetime.now(timezone.utc),
        )

    # ------------------------------------------------------------ reasoning

    @staticmethod
    def _reasoning(level: str, highs: List[str], moderates: List[str], lows: List[str], missing: List[str]) -> str:
        elevated = len(highs) + len(moderates)
        if level == "HIGH":
            opening = (
                "Multiple environmental indicators show elevated risk."
                if elevated >= 2
                else "A single dominant indicator is driving elevated environmental risk."
            )
        elif level == "MODERATE":
            opening = "Environmental indicators show moderate overall risk."
        else:
            opening = "Environmental indicators are largely within acceptable ranges."

        if highs:
            verb = "are the dominant risk factors" if len(highs) > 1 else "is the dominant risk factor"
            tail = [MODERATE_PHRASE[a] for a in moderates] or [LOW_PHRASE[a] for a in lows]
            body = f"{_capitalise(join_and(HIGH_PHRASE[a] for a in highs))} {verb}"
            body += f", while {join_and(tail)}." if tail else "."
        elif moderates:
            verb = "are the main concerns" if len(moderates) > 1 else "is the main concern"
            tail = [LOW_PHRASE[a] for a in lows]
            body = f"{_capitalise(join_and(HIGH_PHRASE[a] for a in moderates))} {verb}"
            body += f", while {join_and(tail)}." if tail else "."
        else:
            body = "No specialist agent reported indicators above guideline thresholds."

        text = f"{opening} {body}"
        if missing:
            text += f" The {join_and(missing)} assessment is unavailable, so confidence is reduced."
        return text

    @staticmethod
    def _insights(payload: CoordinatorInput, highs: List[str], adjustment: float, missing: List[str]) -> List[str]:
        insights: List[str] = []
        if adjustment > 0:
            insights.append(
                f"Correlated high-risk signals from the {join_and(highs)} agents add "
                f"{adjustment * 100:.0f} points to the aggregate score."
            )
        air, water, waste = payload.air, payload.water, payload.waste
        if waste and water and waste.risk_level != "LOW" and water.turbidity is not None and water.turbidity > 5:
            insights.append(
                "Litter density alongside elevated turbidity suggests surface runoff may be "
                "carrying waste into the nearby water body."
            )
        if air and waste and air.risk_level == "HIGH" and waste.counts.plastic >= 8:
            insights.append(
                "High PM2.5 near a plastic-heavy litter zone warrants a check for open waste burning."
            )
        if water and water.sensor_status == "degraded":
            insights.append("Water assessment relies on partial sensor data; one or more probes are not reporting.")
        if missing:
            insights.append(
                f"No report received from the {join_and(missing)} agent; its weight was redistributed."
            )
        return insights

    @staticmethod
    def _data_limitations(payload: CoordinatorInput, missing: List[str]) -> List[str]:
        """Freshness/mock/missing-agent caveats, derived only from the typed specialist reports.

        Deterministic and side-effect free (no I/O, no new imports) — every claim here traces
        back to a field the specialist agent already reported, never an inference or invention.
        """
        limitations: List[str] = []
        air, water, waste = payload.air, payload.water, payload.waste

        if air is not None:
            if air.is_mock:
                limitations.append("Air-quality data is simulated demo data, not a live reading.")
            elif air.data_age_minutes is not None and air.data_age_minutes > 15:
                limitations.append(f"Air-quality readings are {air.data_age_minutes} minutes old.")

        if water is not None:
            if water.is_mock:
                limitations.append("Water-quality data is simulated demo data, not a live reading.")
            elif water.data_age_minutes is not None and water.data_age_minutes > 15:
                limitations.append(f"Water readings are {water.data_age_minutes} minutes old.")
            if water.sensor_status == "degraded":
                limitations.append(f"{water.sensor_name} is reporting partial data; some parameters are missing.")

        if waste is not None:
            if waste.input_type == "upload":
                limitations.append(
                    "Waste analysis is image-based and reflects a single uploaded photo, not continuous monitoring."
                )
            elif waste.is_mock:
                limitations.append("Waste detection is simulated demo data, not a live camera feed.")

        for agent_id in missing:
            limitations.append(f"No {agent_id} assessment is available; treat the overall risk with reduced confidence.")
        return limitations

    @staticmethod
    def _recommendations(payload: CoordinatorInput, overall: float, overall_level: str) -> List[Recommendation]:
        candidates: List[Tuple[float, dict]] = []
        air, water, waste = payload.air, payload.water, payload.waste

        if waste and waste.risk_level != "LOW":
            candidates.append((waste.risk_score * 1.0, dict(
                id="inspect-waste", title="Inspect waste accumulation zone", agent="waste",
                explanation=(f"{waste.total_objects} objects detected at {waste.source_name}, "
                             f"{waste.counts.plastic} of them plastic. Dispatch a sanitation crew and trace the source."),
                action_label="Dispatch crew")))

        if water and water.sensor_status == "degraded":
            candidates.append((max(water.risk_score, 0.5) * 0.95, dict(
                id="restore-sensor", title="Restore water sensor telemetry", agent="water",
                explanation=f"{water.sensor_name} is reporting partial data. Inspect probes and connectivity on {water.sensor_id}.",
                action_label="Open ticket")))
        elif water and water.risk_level != "LOW":
            turbidity = (f"Turbidity at {format_number(water.turbidity)} NTU exceeds the 5 NTU permissible limit"
                         if water.turbidity is not None and water.turbidity > 5
                         else "Water indicators are drifting from safe ranges")
            candidates.append((water.risk_score * 0.95, dict(
                id="investigate-water", title="Investigate nearby water source", agent="water",
                explanation=f"{turbidity} at {water.sensor_name}. Collect a grab sample for lab validation.",
                action_label="Schedule sampling")))

        if overall_level != "LOW":
            candidates.append((overall * 0.7, dict(
                id="increase-monitoring", title="Increase environmental monitoring frequency", agent="all",
                explanation="Raise sampling cadence from 60 to 15 minutes across all agents for the next 24 hours to confirm the trend.",
                action_label="Update cadence")))

        if air and air.risk_level == "HIGH":
            pm = (f"PM2.5 at {format_number(air.pm25)} µg/m³ is {air.pm25 / 15:.1f}× the WHO guideline"
                  if air.pm25 is not None else "Air pollution is at a high-risk level")
            candidates.append((air.risk_score * 0.65, dict(
                id="air-advisory", title="Issue air-quality health advisory", agent="air",
                explanation=f"{pm}. Notify sensitive groups near {air.station_name} and review traffic controls.",
                action_label="Draft advisory")))
        elif air and air.risk_level == "MODERATE":
            candidates.append((air.risk_score * 0.6, dict(
                id="air-watch", title="Monitor air-quality trend", agent="air",
                explanation=f"Air quality at {air.station_name} is moderate. Watch for sustained increases over the next 12 hours.",
                action_label="Set alert")))

        if not candidates:
            candidates.append((0.1, dict(
                id="routine", title="Maintain routine monitoring", agent="all",
                explanation="All indicators are within acceptable limits. Continue the standard sampling schedule.",
                action_label="Acknowledge")))

        candidates.sort(key=lambda item: item[0], reverse=True)
        recommendations = []
        for index, (urgency_score, data) in enumerate(candidates):
            urgency = "Immediate" if index == 0 and urgency_score >= 0.6 else "Within 24h" if urgency_score >= 0.45 else "Routine"
            recommendations.append(Recommendation(priority=index + 1, urgency=urgency, **data))
        return recommendations
